### PR TITLE
Meta-macro-bonanza

### PR DESCRIPTION
**Pull request by martinsvalin**
Use metaprogramming to define GoodTimes functions

    Since GoodTimes requires Meta, we can't use functions in GoodTimes inside Meta.
    This is why Seconds and Now have been moved out, and the @types put in Now.

    `GoodTimes.now` isn't set up by macro, so to make it available straight from
    GoodTimes, there is a delegation to Now. Unfortunately documentation doesn't
    follow along in a delegation, and so it's duplicated.

    Seconds could have been delegated, but then we run into the documentation issue
    again. This time, we can instead define those methods with the macro. It's a
    bit weird, since some of them they essentially duplicate the methods in
    `GoodTimes.Seconds`.

    I've so far been unable to set the parameter names to the macro defined
    functions dynamically. They show up in documentation. Where it would read
    `seconds_ago(seconds)`, it now reads `seconds_ago(integer)`.

    The macro takes a singular version of the time unit. The plural is derived in
    the most naive way: sticking an "s" on the end.

### CODE REVIEW

**Comment by martinsvalin** [comment]
Another take on #5. Not sure this is an improvement on normal functions, either.

File: lib/good_times.ex [hunk 1]
```diff
@@ -26,360 +14,12 @@ defmodule GoodTimes do
...
-  Returns the UTC date and time a day ago.
-
-   ## Examples
-
-      iex> a_day_ago
-      {{2015, 2, 26}, {18, 30, 45}}
-  """
-  @spec a_day_ago :: datetime
-  def a_day_ago, do: days_ago(1)
+  require GoodTimes.Meta
```
Comment (DevL) [inline lib/good_times.ex]: No need to require an imported module as importing a module entails requiring it.

File: lib/meta/meta.ex [hunk 2]
```diff
@@ -0,0 +1,129 @@
+defmodule GoodTimes.Meta do
+  def a_datetime, do: {{2015, 2, 27}, {18, 30, 45}}
```
Comment (DevL) [inline lib/meta/meta.ex]: This looks a bit odd. How about using a module attribute instead?

`@example_datetime {{2015, 2, 27}, {18, 30, 45}}`

**Comment by DevL** [comment]
Very cool and insight-rendering PR!

The most apt description of this approach I can think of is "clever". :grin:

I have a few concerns with number of Modules involved and the fact that the Seconds module really only exists in order to make it easier to generate the examples in the documentation. The naming of the parameters feels like a blocker too, but I'm sure we can find out how to do that by asking the collective wisdom of the Elixir mailing list. :smile:

Now, if we could generate a single module containing all the files, typespecs and docstrings which we then package and publish to Hex I'd be much happier.

Thoughts?

File: lib/now/now.ex [hunk 3]
```diff
@@ -0,0 +1,20 @@
+defmodule GoodTimes.Now do
+  @type year  :: integer
```
Comment (DevL) [inline lib/now/now.ex]: These have changed in master. Using Dialyzer, I found that we were too permissive and changed them accordingly.

**Comment by DevL** [comment]
The more I think about it, the more I'm starting to believe there's a more generic macro or pattern to be extracted from this. Something like a `defderivate` that, based on a supplied base function, generates a function that does something extra with the result of the base function (e.g. in our case multiplies by some factor). Such a generic macro would most likely need to have the `@doc` and potentially some of the `@spec` Mor explicitly passed in, but it might be quite reusable.

We'd need to figure out the naming of arguments to make it work well as otherwise the docs would suffer.

File: lib/meta/meta.ex (hunk 2 above)
Comment (martinsvalin) [inline lib/meta/meta.ex]: I tried that, but got some errors I didn't completely understand. If `@a_datetime` is placed straight in Meta, and I reference it in the quoted documentation part of the macro, I get:

```
lib/good_times.ex:20: warning: undefined module attribute @a_datetime, please remove access to @a_datetime or explicitly set it to nil before access

== Compilation error on file lib/good_times.ex ==
** (FunctionClauseError) no function clause matching in :calendar.datetime_to_gregorian_seconds/1
```

So clearly, `@a_datetime` is nil in this context. I suspect it's looking for it in the module where the macro is expanded, but I don't want to pollute that module. It's not a question of unquoting (the datetime value is not a valid quoted expression, the compiler tells me). I was blocked and simply moved it to a function.

**Comment by martinsvalin** [comment]
I'm actually fine just dropping both of these macro-approaches (#5 and #8). There is beauty in simple functions. Too bad it's easy to get lost in all the doc strings. But that's really more of an editor-deficiency. Maybe we should focus our efforts on getting the sublime-elixir-plugin right, so it lets us fold all doc strings out of our way.

How are the `@doc` and `@spec` directives stored on the module, btw? Is there a way to inspect them programmatically? A general `defderivative` would need to read the `@doc` and `@spec` for a function, and transform them based on an assumption of their structure.

**Comment by DevL** [comment]
Agreed. Dropping #5.

Figuring out a good way to fold away the @doc strings in Sublime Text sounds like a good excuse to tinker a bit with Python. :smile:

I believe that Elixir module attributes normally are only available at compile time, but clearly @doc and @spec must be exceptions as the information can be obtained using the iex helper functions `h` and `s` respectively. If nothing else, we could dive into the iex source code in an attempt to find out how to access their data.

**Comment by martinsvalin** [comment]
I'm closing this PR as well. I like how it shows up front what time units it supports, and cuts down on duplication, but there's still quite a bit of duplication between the different functions for one time unit, so even there it doesn't really live up to its purpose. And the reduced duplication comes at the cost of spreading functions among modules, and generally reducing readability.

Maybe we'll revisit after mastering metaprogramming in Elixir. ;) This was really my first taste of doing some of my own.

**Comment by martinsvalin** [comment]
About programmatically finding a `@doc`:

```
iex> Code.get_docs(GoodTimes, :docs) \
...>   |> List.keyfind({:weeks_before, 2}, 0) \
...>   |> elem(4) \
...>   |> IO.puts
Returns the UTC date and time the specified weeks before the given datetime.

## Examples

    iex> 3 |> weeks_before {{2015, 2, 27}, {18, 30, 45}}
    {{2015, 2, 6}, {18, 30, 45}}

:ok
```

**Pull request closed without merging**